sync/optimistic.md

```diff
@@ -81,10 +81,29 @@ def is_execution_block(block: BeaconBlock) -> bool:
 
 ```python
 def is_optimistic_candidate_block(opt_store: OptimisticStore, current_slot: Slot, block: BeaconBlock) -> bool:
-    justified_root = opt_store.block_states[opt_store.head_block_root].current_justified_checkpoint.root
+    state = opt_store.block_states[opt_store.head_block_root]
+    justified_root = state.current_justified_checkpoint.root
+    finalized_root = state.finalized_checkpoint.root
     justified_is_execution_block = is_execution_block(opt_store.blocks[justified_root])
     block_is_deep = block.slot + SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY <= current_slot
-    return justified_is_execution_block or block_is_deep
+    return justified_is_execution_block or block_is_deep or has_verified_ancestor(opt_store, block, finalized_root)
+```
+
+```python
+def has_verified_ancestor(opt_store: OptimisticStore, block: BeaconBlock, finalized_root: Hash32) -> bool:
+    while True:
+        if not is_execution_block(block):
+            return False
+        elif hash_tree_root(block) not in opt_store.optimistic_roots:
+            return True
+        # Implementations MAY choose to remove this check and iterate further
+        # back in the chain.
+        elif hash_tree_root(block) == finalized_root:
+            return False
+        elif block.parent_root == Hash32() or block.parent_root not in opt_store.blocks:
```

Comment:
Won't just `block.parent_root not in opt_store.blocks` be sufficient in this case?